okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java

```diff
@@ -600,6 +600,32 @@ public void run() {
         // ClientFrameHandler need to be started after connectionPreface / settings, otherwise it
         // may send goAway immediately.
         executor.execute(clientFrameHandler);
```

Comment:
@ejona86 Just to have a clear understanding, I thought this is the reading thread that runs for the life of the transport like you said. If we happen to create a new thread and terminate it before this reading thread even starts then how are we verifying that there are atleast two threads? I'm a bit confused here regarding serializingExecutor and executor during handshake.

Reply:
This is late in the transport's lifetime to check if resources are available. We've done TLS and such with the server. We want to check that resources are available before we get to this point.

The `lock.notify()` thread is running very briefly. It'd be under a millisecond. If there's 10 transports, you only need 11 threads here. But we really need up to 20 threads for 10 transports.

The runnable on 504 is using the same thread as this runnable (serializing executor makes sure they don't run concurrently). Checking for a second thread at the very beginning of the handshake prevents doing I/O only to throw away the transport (because a second thread was unavailable). Keeping the second thread around for the handshake makes it more likely to starve _some other transport_ such that it fails (and frees up a thread).